tortoise/algorithm.go

```diff
@@ -0,0 +1,212 @@
+	GlobalThreshold *big.Rat      `mapstructure:"tortoise-global-threshold"` // threshold for finalizing blocks and layers
+	LocalThreshold  *big.Rat      `mapstructure:"tortoise-local-threshold"`  // threshold for choosing when to use weak coin
+	RerunInterval   time.Duration `mapstructure:"tortoise-rerun-interval"`
+	MaxExceptions   int           `mapstructure:"tortoise-max-exceptions"` // if candidate for base block has more than max exceptions it will be ignored
+
+	LayerSize                uint32
+	BadBeaconVoteDelayLayers uint32 // number of layers to delay votes for blocks with bad beacon values during self-healing
+}
+
+// DefaultConfig for Tortoise.
+func DefaultConfig() Config {
+	return Config{
+		LayerSize:                30,
+		Hdist:                    10,
+		Zdist:                    8,
+		ConfidenceParam:          2,
+		WindowSize:               100,
+		GlobalThreshold:          big.NewRat(60, 100),
+		LocalThreshold:           big.NewRat(20, 100),
+		RerunInterval:            24 * time.Hour,
+		BadBeaconVoteDelayLayers: 6,
+		MaxExceptions:            30 * 100, // 100 layers of average size
+	}
+}
+
+// Tortoise is a thread safe verifying tortoise wrapper, it just locks all actions.
+type Tortoise struct {
+	logger log.Log
+	ctx    context.Context
+	cfg    Config
+
+	eg     errgroup.Group
+	cancel context.CancelFunc
+
+	mu  sync.RWMutex
+	org *organizer.Organizer
+
+	// update will be set to non-nil after rerun completes, and must be set to nil once
+	// used to replace trtl.
+	update *rerunResult
+	// persistMu is needed to allow concurrent BaseBlock and Persist call
+	// but to prevent multiple concurrent Persist calls
+	persistMu sync.Mutex
+	trtl      *turtle
+}
+
+// Opt for configuring tortoise.
+type Opt func(t *Tortoise)
+
+// WithLogger defines logger for tortoise.
+func WithLogger(logger log.Log) Opt {
+	return func(t *Tortoise) {
+		t.logger = logger
+	}
+}
+
+// WithContext defines context for tortoise.
+func WithContext(ctx context.Context) Opt {
+	return func(t *Tortoise) {
+		t.ctx = ctx
+	}
+}
+
+// WithConfig defines protocol parameters.
+func WithConfig(cfg Config) Opt {
+	return func(t *Tortoise) {
+		t.cfg = cfg
+	}
+}
+
+// New creates Tortoise instance.
+func New(db database.Database, mdb blockDataProvider, atxdb atxDataProvider, beacons blocks.BeaconGetter, opts ...Opt) *Tortoise {
+	t := &Tortoise{
+		ctx:    context.Background(),
+		logger: log.NewNop(),
+		cfg:    DefaultConfig(),
+	}
+	for _, opt := range opts {
+		opt(t)
+	}
+
+	if t.cfg.Hdist < t.cfg.Zdist {
+		t.logger.With().Panic("hdist must be >= zdist",
+			log.Uint32("hdist", t.cfg.Hdist),
+			log.Uint32("zdist", t.cfg.Zdist),
+		)
+	}
+	t.trtl = newTurtle(
+		t.logger,
+		db,
+		mdb,
+		atxdb,
+		beacons,
+		t.cfg,
+	)
+
+	if err := t.trtl.Recover(); err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			t.trtl.init(t.ctx, mesh.GenesisLayer())
+		} else {
+			t.logger.With().Panic("can't recover turtle state", log.Err(err))
+		}
+	}
```

Comment:
What do you think about removing the `else` statement?
```suggestion
	if err := t.trtl.Recover(); err != nil {
		if !errors.Is(err, database.ErrNotFound) {
			t.logger.With().Panic("can't recover turtle state", log.Err(err))
		}
		t.trtl.init(t.ctx, mesh.GenesisLayer())
	}
```